cmd/dockerd/daemon.go

```diff
@@ -450,23 +450,21 @@ func newRouterOptions(ctx context.Context, config *config.Config, d *daemon.Daem
 			Named:    d.ImageNamedByBuildkit,
 		},
 	})
-	if err != nil {
-		return routerOptions{}, err
-	}
 
 	bb, err := buildbackend.NewBackend(d.ImageService(), manager, bk, d.EventsService)
 	if err != nil {
-		return routerOptions{}, errors.Wrap(err, "failed to create buildmanager")
+		return nil, nil, errors.Wrap(err, "failed to create buildmanager")
 	}
+	return bb, sm, nil
+}
 
-	return routerOptions{
-		sessionManager: sm,
-		buildBackend:   bb,
-		features:       d.Features,
-		buildkit:       bk,
```

Comment:
I think this may be related to the failure; before the refactor, `routerOptions.buildkit` was set, but in the new code we no longer do this.

And it's called during daemon shutdown to close buildkit; https://github.com/moby/moby/blob/2f6953fcd8a411816ea84f4d24fd9ea101c19ba7/cmd/dockerd/daemon.go#L353-L355


Perhaps `initBuildkit` should return a closer function instead?

Reply:
Yeah, bad copy/paste in the refactor from the prior change (which set buildkit after initializing the struct).